### PR TITLE
Api token updates

### PR DESCRIPTION
**Pull request by SchrodingersGat**
Fixes https://github.com/inventree/InvenTree/issues/2326

This PR implements a custom extension of the current API token implementation:

- Adds custom expiry to the token
- Adds a "name" to each token

### TODO

- [x] Add new model
- [x] Migrate existing tokens
- [x] Admin integration
- [x] Update auth to use new model (and check if the token is currently active)
- [x] Implement / extend method for creating tokens "on the fly" via the API (or do we remove this entirely)
- [ ] Look into integration with the mobile app
- [x] Look into integration with the python API
- [x] Unit tests

A key point here is to preserve backwards compatibility with any existing external systems (such as the mobile app) which use tokens and expect them to appear in a certain way on the API.

### References

- https://github.com/inventree/inventree-app/issues/351
- https://idiomaticprogrammers.com/post/how-to-implement-auto-expiring-token-in-django-rest-framework/

### CODE REVIEW

File: InvenTree/users/migrations/0008_apitoken.py
Comment (matmair) [inline InvenTree/users/migrations/0008_apitoken.py:1]: Changing an existing migration is probably not what we want to do.

Comment (matmair) [inline InvenTree/users/migrations/0008_apitoken.py:1]: Changing an existing migration is probably not what we want to do.

**Review by matmair — COMMENTED** [review]
I think this would be a good point to address some things that are not optimal about the current design of tokens and change things to be more like best-practice examples like GitHub and stripe.
- No token without expiry
- Configurable max expiry (industry standard seems to be 1 year nowadays)
- Token events (creation, deletion, expert imminent)
- Use an identifiable token format (see https://github.blog/2021-04-05-behind-githubs-new-authentication-token-formats/#identifiable-prefixes) - gives us the option to support multiple token formats in the future
- Add metadata support to the model and issuing API (ie. the app could start sending the model and name of the phone when requesting a token to make identifying easier)

**Comment by matmair** [comment]
@wolflu05 it is not really easy as the tokens currently are basically impersonating their user so we would probably need separate users or change all permission checks (which are user based currently)

File: InvenTree/users/migrations/0008_apitoken.py
Comment (SchrodingersGat) [reply InvenTree/users/migrations/0008_apitoken.py:1]: This is not an *existing* migration though, it is new as part of this PR

Comment (matmair) [reply InvenTree/users/migrations/0008_apitoken.py:1]: Thats funny, the UI on the iOS app showed it in yellow. In vs-code it is shown as new.

Comment (SchrodingersGat) [reply InvenTree/users/migrations/0008_apitoken.py:1]: Probably yellow because I created it in one commit, and edited in a successive commit

File: InvenTree/users/api.py [hunk 1]
```diff
@@ -187,15 +186,32 @@ class GetAuthToken(APIView):
...
         if request.user.is_authenticated:
-            # Get the user token (or create one if it does not exist)
-            token, created = Token.objects.get_or_create(user=request.user)
-            return Response({
+
+            user = request.user
+            name = request.query_params.get('name', '')
+
+            # Delete any matching tokens
+            ApiToken.objects.filter(user=user, name=name).delete()
```
Comment (matmair) [inline InvenTree/users/api.py:200]: Comment: I think it would be prudent to log it a token was actually deleted for development (maybe log the identifier)

File: InvenTree/users/admin.py [hunk 2]
```diff
@@ -17,9 +17,9 @@
 class ApiTokenAdmin(admin.ModelAdmin):
     """Admin class for the ApiToken model."""
 
-    list_display = ('key', 'user', 'name', 'expiry', 'active')
-    fields = ('user', 'name', 'revoked', 'expiry')
-    readonly_fields = ('key', 'created')
+    list_display = ('token', 'user', 'name', 'expiry', 'active')
+    fields = ('token', 'name', 'revoked', 'expiry')
+    readonly_fields = ('token', 'created')
```
Comment (matmair) [inline InvenTree/users/admin.py:22]: Question: Why is that changed? The key is normally used for identification in systems like this

**Review by matmair — APPROVED** [review]

File: InvenTree/users/api.py (hunk 1 above)
Comment (SchrodingersGat) [reply InvenTree/users/api.py:200]: Good call I'll add that.

File: InvenTree/users/admin.py (hunk 2 above)
Comment (SchrodingersGat) [reply InvenTree/users/admin.py:22]: The "token" is the redacted version, once it is created the raw value is not shown in the admin interface - I believe this is what you requested?
